azure/Guidelines.md

```diff
@@ -540,6 +546,18 @@ Below is an example of JSON for a Rectangle and Circle:
 ```
 Both Rectangle and Circle have common fields: `kind`, `fillColor`, `lineColor`, and `subscription`. A Rectangle also has `x`, `y`, `width`, and `length` while a Circle has `x`, `y`, and `radius`. The `subscription` is a nested polymorphic type. A `free` subscription has no additional fields and a `paid` subscription has `expiration` and `invoice` fields.
 
+<a name="json-polymorphism-kind-extensible"></a>
+:ballot_box_with_check: **YOU SHOULD** define the kind field of a polymorphic type to be an extensible enum.
+
+<a name="json-polymorphism-kind-immutable"></a>
+:warning: **YOU SHOULD NOT** allow an update (patch) to change the kind field of a polymorphic type.
+
+<a name="json-polymorphism-versioning"></a>
+:warning: **YOU SHOULD NOT** return properties of a polymorphic type that are not defined for the api-version specified in the request.
+
+<a name="json-polymorphism-arrays"></a>
+:warning: **YOU SHOULD NOT** specify a polymorphic type as the type of an array property.
```

Comment:
This is a special case of the known problem that arrays are not "merge-patch-friendly" -- to add or remove an element of an array with merge-patch you must specify the entire array contents, modified accordingly. This is problematic for polymorphic types because it acts like a "replace" on individual elements, so properties of the element that were not returned due to older api-version get lost.

Reply:
Oh yea, thanks for the refresher. Maybe rephrase this bullet:
**YOU SHOULD NOT** have a property whose value is an array of polymorphic types.

Reply:
So what should you do if you return a sequence of values? For example, Form Recognizer returns a set of recognized fields. They are of different type.

The json-patch gotcha only applies to updatable resources.

Reply:
Good point. This rule should only apply to updatable resources.